elastic/kibana
# mock `elastic-apm-node` in `@kbn/test` jest preset

## Summary

Fix https://github.com/elastic/kibana/issues/117255

In https://github.com/elastic/kibana/issues/117255 we investigated a memory leak in jest environment, that lead to `elastic-apm-node` patching native modules at runtime (please refer to the issue for more details).

This PR fixes it by mocking `elastic-apm-node` in our `@kbn/test` preset

Thanks to @FrankHassanabad and @academo for providing the scripts helping identifying the leak and starting the investigation.

### How to test

You need to run tests that are identified as leaking using `--runInBand`, and confirm that no memory leaks with this PR's branch.

I created a `x-pack/plugins/alerting/server/sample_test.ts` file with the following test:


```ts
import { plugin } from '../../actions/server';

describe('identifying the leak', () => {
  test('is hard', () => {
    Object.keys(plugin);
  });
});
```

and used the following script (credits to @FrankHassanabad):

```sh
#!/bin/sh

# Set your kibana home here
KIBANA_HOME=/path/to/elastic/kibana
# Set your kibana project here
KIBANA_PROJECT=x-pack/plugins/alerting/jest.config.js
# When it does leak set this to around 2 gigs to get memory snapshots without issues.
MAX_MEMORY=500
# Set this to true to activate "--inspect-brk" below
CHROME_INSPECT=false

# Useage if no arguments
if [ $# -eq 0 ]; then
  echo "Usage is ./copy_tests.sh <test name>"
  exit 1
fi

# Check if file exists or not
if [[ ! -f $1 ]]; then
  echo "File does not exist"
  exit 1
fi

# Gets the directory of your first argument
DIR="$(dirname "${1}")"

# Remove all the copied scripts even if cntrl-C is pressed or other issues.
trap "rm $DIR/script_copy_test_*.test.ts" EXIT

# Loops max times and copy the single test in place
for i in {1..100}
do
  cp $1 $DIR/script_copy_test_$i.test.ts
done

pushd $KIBANA_HOME;
if [ "$CHROME_INSPECT" = true ] ;
then
  node --max-old-space-size=$MAX_MEMORY --inspect-brk --expose-gc ./node_modules/.bin/jest --verbose --runInBand --logHeapUsage --detectOpenHandles --no-cache --config $KIBANA_PROJECT $DIR/script_copy_test_*.ts;
else
  node --max-old-space-size=$MAX_MEMORY --expose-gc ./node_modules/.bin/jest --verbose --runInBand --logHeapUsage --detectOpenHandles --no-cache --config $KIBANA_PROJECT $DIR/script_copy_test_*.ts;
fi
popd;
```

Using with

```sh
./memory_leak.sh x-pack/plugins/alerting/server/sample_test.ts
```
